qiskit/transpiler/passes/graysynth.py

```diff
@@ -0,0 +1,250 @@
+            condition = True
+            while condition:
+                condition = False
+                for j in range(number):
+                    if (j != qubit) and (sum(cnots[j]) == len(cnots[j])):
+                        condition = True
+                        qcir.cx(qreg[j], qreg[qubit])
+                        state[qubit] ^= state[j]
+                        index = 0
+                        for icnots in cnots_copy:
+                            if np.array_equal(icnots, state[qubit]):
+                                qcir.t(qreg[qubit])
+                                cnots_copy = np.delete(cnots_copy, index, axis=0)
+                                if index == len(cnots_copy):
+                                    break
+                                index -= 1
+                            index += 1
+                        for x in remove_duplicates(sta.elements() + [[cnots, ilist, qubit]]):
+                            [cnotsp, _, _] = x
+                            if cnotsp == []:
+                                continue
+                            for ttt in range(len(cnotsp[j])):
+                                cnotsp[j][ttt] ^= cnotsp[qubit][ttt]
+        if ilist == []:
+            continue
+        # See line 18 in pseudo-code of Algorithm 1 in the aforementioned paper
+        j = ilist[np.argmax([[max(row.count(0), row.count(1)) for row in cnots][k] for k in ilist])]
+        cnots0 = []
+        cnots1 = []
+        for y in list(map(list, zip(*cnots))):
+            if y[j] == 0:
+                cnots0.append(y)
+            elif y[j] == 1:
+                cnots1.append(y)
+        cnots0 = list(map(list, zip(*cnots0)))
+        cnots1 = list(map(list, zip(*cnots1)))
+        if qubit == epsilon:
+            sta.push([cnots1, list(set(ilist).difference([j])), j])
+        else:
+            sta.push([cnots1, list(set(ilist).difference([j])), qubit])
+        sta.push([cnots0, list(set(ilist).difference([j])), qubit])
+    qcir = cnot_synth(qcir, state, qreg, number, nsections)
+    return qcir
+
+
+def cnot_synth(qcir, state, qreg, number, nsections):
+    """
+    This function is an implementation of the algorithm for optimal synthesis of linear
+    reversible circuits, as described in the following paper:
+    "Optimal synthesis of linear reversible circuits."
+    Patel, Ketan N., Igor L. Markov, and John P. Hayes.
+    Quantum Information & Computation 8.3 (2008): 282-294.
+
+    Args:
+        qcir (QuantumCircuit): the initial Quantum Circuit
+        state (numpy.matrix): n by n matrix, describing the state of the input circuit
+        qreg (QuantumRegister): a Quantum Register
+        number (int): the number of quantum bits in the circuit
+        nsections (int): the number of partitions used in the below algorithm
+
+    Returns:
+        QuantumCircuit: a Quantum Circuit with added C-NOT gates
+    """
+
+    state = np.matrix(state)  # Making sure that state is a numpy matrix
```

Comment:
the docstring says that `state` is expected to be a numpy.matrix. So do we still need this line? instead I think we should check and throw a meaningful error if it is not. this conversion will fail if `state` is not a valid input to `np.matrix()`.

Reply:
Ok, I changed it